pfs_middleware/pfs_middleware/middleware.py

```diff
@@ -855,6 +863,90 @@ def _fetch_appropriate_acl(self, ctx):
+                return swob.HTTPLengthRequired(request=req)
+            json_payloads = req.body_file.read(MAX_RPC_BODY_SIZE).split(b'\n')
+            if req.body_file.read(1):
+                return swob.HTTPRequestEntityTooLarge(request=req)
+        else:
+            if cl > MAX_RPC_BODY_SIZE:
+                return swob.HTTPRequestEntityTooLarge(request=req)
+            json_payloads = req.body.split(b'\n')
+
+        if self.bypass_mode == 'read-write':
+            allowed_methods = rpc.allow_read_write
+        else:
+            allowed_methods = rpc.allow_read_only
+
+        payloads = []
+        for i, json_payload in enumerate(x for x in json_payloads
+                                         if x.strip()):
+            try:
+                payload = json.loads(json_payload)
+                if payload['jsonrpc'] != '2.0':
+                    raise ValueError
+                if isinstance(payload['method'], string_types):
+                    raise ValueError
+                if payload['method'] not in allowed_methods:
+                    raise ValueError
+                if not (isinstance(payload['params'], list) and
+                        len(list) == 1 and
+                        isinstance(payload['params'][0], dict)):
+                    raise ValueError
+            except (TypeError, KeyError, ValueError):
+                self.logger.debug('Error parsing payload #%d %r',
+                                  i, json_payload, exc_info=True)
+                return swob.HTTPBadRequest(
+                    request=req,
+                    body=('Could not parse/validate JSON payload %r' %
+                          (json_payload, )))
+            payloads.append(payload)
+
+        # TODO: consider allowing more than one payload per request
+        if len(payloads) != 1:
+            return swob.HTTPBadRequest(
+                request=req,
+                body='Expected exactly one JSON payload')
+
+        # Our basic validation is done; spin up a connection and send requests
+        client = utils.JsonRpcClient(ctx.proxyfsd_addrinfo)
+        payload = payloads[0]
+        try:
+            if 'id' not in payload:
+                payload['id'] = str(uuid.uuid4())
+            payload['params'][0]['AccountName'] = ctx.account_name
+            response = client(payload, self.proxyfsd_rpc_timeout,
+                              raise_on_rpc_error=False)
+        except eventlet.Timeout:
+            self.logger.debug(
+                "Timeout (%.6fs) communicating with %s, calling %s",
+                self.proxyfsd_rpc_timeout, ctx.proxyfsd_addrinfo,
+                payloads[0]['method'])
+            return swob.HTTPBadGateway(request=req)
+        except socket.error as err:
+            self.logger.debug("Error communicating with %r: %s.",
+                              ctx.proxyfsd_addrinfo, err)
+            return swob.HTTPBadGateway(request=req)
+        else:
+            if 'error' in response:
```

Comment:
I don't think I've got a simple way to *NOt* include the error field in a response.
I realize the JSON RPC 2.0 spec says something like "only error or payload" can be included. The Go JSON Marshaler doesn't make that very easy. It's possible... but we never did that in the package jrpcfs layer. I *could* go re-write all of that... but don't want to do that now.

Indeed, I'd much sooner re-write it by ditching JSON RPC altogether. It's crap imho.

Reply:
I guess what I am saying is that an "empty" error in the response should also be allowed.

Heck, for all I know, the JSON RPC layer we are leveraging takes care of this though. Do you know?

Reply:
I *think* it might be OK? I thought I cribbed this pattern from somewhere... but no, it looks like we've got

            errstr = result.get("error")
            if errstr:
                errno = utils.extract_errno(errstr)
                raise utils.RpcError(errno, errstr)

down in `_rpc_call`. I'm fine with changing it to `if response.get('error'):`. #willfix

Reply:
I'm sure you're right... in thinking about it, that funky JSON RPC layer we are using in ProxyFS does all sorts of stuff for us. It codes the event loop, serializes/deserializes, etc... all we do is implement a bunch of functions that, for e.g. Server.RpcFooRequest/Reply, we make a public function named Foo() that takes two by-reference args: in *FooRequest struct, and reply *FooReply struct. That's it... so this must be ok...